Reader: Add new post detail header

dvdchr opened the pull request:
Refs #21644 

This adds the new post detail header, now in SwiftUI.

Before | After
-|-
![before_light](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/b5a8d153-f331-4144-8f29-efde9caf6936) | ![after_light](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/0857008d-6657-4b76-a401-6cfb9ac87e61)
![before_dark](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/3d49ad67-6885-4291-ac08-63cb74f44288) | ![after_dark](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/1cf8c7fe-8983-4b43-9bcf-a45a7294ef58)

In the code, I've found that there's a thing called an "attribution view" in Reader detail. I haven't figured out what it's used for, or what kind of posts display this view. Will try to figure this out later. 

Furthermore, I've practically recreated the legacy Follow button in SwiftUI, and turns out that in the Reader detail, the Follow button uses a different style for iPhone and iPad. This has been recreated in SwiftUI, although not 100% perfect. I think it should be okay since we will update the button again soon.

I've decided to break my large changes into 3 so it's more manageable to review and test. Some things are not included in this PR:

1. The Follow button is still using the legacy design.
2. The tags list is moved but still uses the legacy design.
3. Some accessibility/VoiceOver polish.
4. Action tab bar.
5. Sans-serif font for the body. Turns out this is quite trivial to do, so we'll be able to ship this in the iteration!

Note: there's also a weird legacy issue where the featured image overlaps into the header area by about 4px. I haven't figured out why this happened, and for now, I've added around 4px additional top padding to the new header. The downside is, this will mean 4px extra padding when the post does not have a featured image.

## To test

- Launch the Jetpack app.
- Turn on the `readerImprovements` feature flag.
- Navigate to Reader.
- Open any Reader post — pick two: one with feature image, and one without.
- 🔎 Verify that the new header design is shown.
- 🔎 Verify that the Follow/Unfollow button works.
- 🔎 Verify that tapping the blog name navigates you to the Site stream.
- Select a post that contains several tags, such that they are collapsed. 
- Tap on the "+" button in the tags list to expand the tags.
- 🔎 Verify that expanding/collapsing the tags works correctly.
- Rotate the device to landscape orientation.
- 🔎 Verify that the layout doesn't break.

Additional variations:
- Test in iPad.
- Test between light/dark schemes.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The feature flag gates all the new codes.

6. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

7. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Initial new header view

dvdchr committed:
Feed the view model with actual data

dvdchr commented:
Encountered a CI error:

```
Removing Registered VM xcode-15.0
Error: ShellOut encountered an error
Status code: 255
Message: "Failed to unregister the VM: Unable to perform the action because the virtual machine is busy. The virtual machine is currently running. Please try again later."
```

Will give it a bit of time before retrying again. Looks like all the workers are busy.

osullivanchris reviewed (approved):
Nice work!

A couple of small tweaks I think we could improve when I see the designs live like this:
- a 1pt white/black border (light/dark mode) around the smaller author image to clearly separate the two images that are nested together
- if possible nudge the smaller avatar right 2pt and down 2pt to be slightly outside the larger circle

Another thought looking at it live, which maybe we can't address:
- in discover, the lack of avatars is a bummer. I believe we have raised that with the web team as its an API issue.
- sometimes sites have the same site name and author name. In this case it would be cool to ignore author and only show the site, rather than duplicating. But I don't know if thats possible to identify. It might just be that our users have used the same site name and username.

dvdchr commented:
Thanks for the review @osullivanchris!

> a 1pt white/black border (light/dark mode)
> nudge the smaller avatar right 2pt and down 2pt

Should be possible. Will update this PR soon!

> In this case it would be cool to ignore author and only show the site, rather than duplicating. But I don't know if thats possible to identify.

On the client side, I guess it should be possible to do a quick text comparison and hide the author's name if both values are the same. 

---

Oh, there's one permission-related change in this PR. You need a WP.com account to follow a site, but in the production version, the Follow button is shown in Reader Detail. Tapping on this button will always show the "Unable to follow" snack bar.

In this PR, I've properly removed the button if the user is not a WP.com account. Although, this could be an opportunity for us to "upsell" non-WP.com users to create an account?

osullivanchris commented:
> Should be possible. Will update this PR soon!

great stuff!

> On the client side, I guess it should be possible to do a quick text comparison and hide the author's name if both values are the same.

if this is a reasonable solution it could be nice.

> In this PR, I've properly removed the button if the user is not a WP.com account. Although, this could be an opportunity for us to "upsell" non-WP.com users to create an account?

I like this approach more! I think keep the existing behaviour and maybe we can iterate later to suggest sign up from that snackbar?

wargcm commented on WordPress/Classes/ViewRelated/Reader/Detail/Views/ReaderDetailNewHeaderView.swift:
```diff
@@ -0,0 +1,421 @@
+
+    func refreshFollowButton() {
+        viewModel.isFollowingSite = true
+    }
```
Will this be updated in a future PR to match the current state? Or is it being set to `true` here, correct?

dvdchr replied:
No, it was a mistake. Thanks for catching this! It's fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21674/commits/1f638251055c517d0977a83fe00e7eebb795ab4d.

wargcm commented on WordPress/Classes/ViewRelated/Reader/Detail/Views/ReaderDetailNewHeaderView.swift:
```diff
@@ -0,0 +1,421 @@
+        translatesAutoresizingMaskIntoConstraints = false
+
+        let headerView = ReaderDetailNewHeaderView(viewModel: viewModel) { [weak self] in
+            guard let swiftUIView = self?.subviews.first else {
```
This causes a warning since the variable is unused.
```suggestion
            guard self?.subviews.first != nil else {
```

dvdchr replied:
Thanks! This is fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21674/commits/4623b4da0d404e893acebb61bf4bbfb360cc42db.

wargcm commented on WordPress/Classes/ViewRelated/Reader/Detail/Views/ReaderDetailNewHeaderView.swift:
```diff
@@ -0,0 +1,421 @@
+    private let isLoggedIn: Bool
+
+    // TODO: Populate this with values from the ReaderPost.
+    private lazy var viewModel: ReaderDetailHeaderViewModel = {
+        $0.topicDelegate = self
+        $0.headerDelegate = delegate
+        return $0
+    }(ReaderDetailHeaderViewModel())
```
On my end, this gets initialized before the `delegate` for `ReaderDetailNewHeaderViewHost` is set. Which causes all the delegate functions in the view model to fail because the `headerDelegate` is `nil`.

dvdchr replied:
Thanks, Chris. Yeah, I think I messed up my commits. I stumbled upon this exact problem, and I've already fixed it... but it didn't get included in this PR. This is now fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21674/commits/1bce13ec0b5c746294c6b50e534acb674953508c.

wargcm reviewed (commented):
⚠️ I was unable to verify:
- That the Follow/Unfollow button works.
- That tapping the blog name navigates you to the Site stream.

✅ I did verify:
- New header design is shown.
- Expanding/collapsing the tags works correctly.
- Layout doesn't break on device rotation.

dvdchr committed:
Add borders to author avatar image

dvdchr committed:
Fix Follow button functionality

dvdchr committed:
Hide author's name if it matches site name

dvdchr committed:
Merge branch 'trunk' into feature/reader-detail-new-header

dvdchr committed:
Always show follow button

dvdchr commented:
Pushed some updates (cc @osullivanchris):
- Added a border to the author's avatar image.
- The author's name is now hidden (along with the `•` separator)  if it exactly matches the site title.
- The Follow button is now always shown.

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/d707b625-a98d-467c-996f-b2c18bc425e1" width=350>


I've also fixed the issues highlighted in the code review. Ready for your second round of review please @wargcm, thanks!

osullivanchris commented:
@dvdchr looks good thanks for those fixes 👍 

What is the max number of lines for a site title in that row? I am half thinking we should just cap it at one line. It breaks the alignment of the avatar, text and follow button a bit when the site title is more than one line. If someone wants to see the full site title, they can tap through to the full site. Does that seem like a bad idea?

osullivanchris commented:
Sorry one more thing I didn't account for. If the avatars are mostly white or black, their shape gets lost. They need some kind of border. This is a bit tricky with the lock up with two icons. But starting with the large avatar, adding a 0.5pt black/white border @ 10% opacity solves it. You can see a white avatar now has a visible shape.

<img width="1818" alt="Screenshot 2023-10-05 at 10 30 52" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/f6a30b17-3d71-4cf9-b391-6aadd429eb5e">

If its possible, it works much better to place the border inside/overlayed on the image, rather than outside. The border takes less attention then, when the image is not white/black, as shown below.

<img width="1972" alt="Screenshot 2023-10-05 at 10 31 49" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/12250839-7f29-4e9f-85dd-796817c3815f">

The really gnarly one is making the 2 image lockup work well with borders. I added that white/black border, which creates a nice cut out effect. To preserve that for most images, while making the white/black images have a visible border, I needed to do 2 borders on the small avatar. First there is a 0.5pt black @ 10% border inside the image. This does the job for the white image of showing its shape. Then there is the 1pt fully white border we already added, which gives the separation between large and small circle and gives that cut out effect. 

<img width="1556" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/d0e1a3c8-38ef-4b68-80c8-a90a99440b6d">
<img width="1634" alt="Screenshot 2023-10-05 at 10 36 17" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/40aa70ae-732c-4252-b90e-a98206d72de0">


If this is impossible or driving you crazy, the simpler thing to do would be to only have one border on the small avatar. This would be the subtle 0.5px border @ 10% black. However, we lose the nice cut out effect and it doesn't look great in the majority case where the image is not white (right)

<img width="1541" alt="Screenshot 2023-10-05 at 10 37 21" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/d3f23874-2666-483c-b867-aa25ce48fd91">

dvdchr commented:
@osullivanchris 

> What is the max number of lines for a site title in that row?

Currently, it is not set, so it is multiline. The avatar is set to be fixed and not scale along with the text container, so it'll sit at the center vertically when dealing with multilines. Here's how it looks:

![Simulator Screenshot - iPhone 15 - 2023-10-05 at 18 07 05](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/112e096a-9d10-4fd3-9c52-ba1140836b8a)

> I am half thinking we should just cap it at one line.

Sounds good to me, but what about the author name? In case of a really long name, I'd guess we wanna truncate the name instead of the timestamp?

```
Site Title
Some really long... • 5 days ago
```

---

> If the avatars are mostly white or black, their shape gets lost. They need some kind of border.
> ... adding a 0.5pt black/white border @ 10% opacity solves it

Love the details, and this should be doable! One quick question about the inner 0.5pt border—do we use the black border in a light scheme and the light border in a dark scheme?

osullivanchris commented:
@dvdchr lets cap the site title to 1 line. And also the Author name. For the author's name, what you suggested is good yep.

> Love the details, and this should be doable! One quick question about the inner 0.5pt border—do we use the black border in a light scheme and the light border in a dark scheme?

took me a bit of thinking but yes! that's right. Both borders flip their colors when in dark mode.

<img width="1488" alt="Screenshot 2023-10-05 at 14 03 20" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/f0510b0a-7911-4151-bb1b-63b254f9705f">

dvdchr committed:
Add an inward border to both circular avatar views

dvdchr committed:
Limit site title and author name to one line

dvdchr commented:
Hey @osullivanchris! Here's a fresh set of screenshots for ya 😄 

### One-liner update

The site title and author's name are now capped to one line! ...although, note that this screenshot was taken before the border changes.

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/a4c0277d-f98f-49a9-935a-f018d83e2a98" width=350>

### Border updates

I've temporarily edited the avatars to display black on dark background or white on light background, to highlight the inward border. The border feels too faint in the dark scheme, but perhaps this was intended?

Light | Dark
-|-
![Simulator Screenshot - iPhone 15 - 2023-10-05 at 23 00 44](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/97b583dd-55fd-4eb0-bfaa-2fbf24e87d13) | ![Simulator Screenshot - iPhone 15 - 2023-10-05 at 23 01 01](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/746dd829-3c33-4300-a917-6caa5bd7976f)

osullivanchris commented:
@dvdchr yep agree let's double it to 20% white or whatever value you find works.

Truncated site / author names look good

dvdchr committed:
Bumped inner border opacity to 0.2 for dark scheme

wargcm reviewed (approved):
LGTM! 👍 

I did find one issue while re-testing, but I'd say it's an edge case. So I'll leave it up to you if you want to fix it in this PR or a future one.

https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/a0a17334-f252-48d6-9079-6c744a3260d0

dvdchr commented:
> I did find one issue while re-testing, but I'd say it's an edge case. 

Thanks @wargcm. I am separately working on updating the tags list, so I'll address the issue there!
